### ISSUE TITLE
Restore (and test more reliabley) twisted.python.log's capture of warnings

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6087]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6087|
|Type|release blocker: regression|
|Created|2012-10-09 15:07:59Z|
|Branch|https://github.com/twisted/twisted/tree/warnings-6087|

Warnings are supposed to end up in the Twisted log once logging is initialized, but [[#5885](https://github.com/twisted/twisted/issues/5885)](https://github.com/twisted/twisted/issues/5885) broke that functionality. It should be fixed (reverting #5885 would be somewhat painful at this point) and the tests improved so this doesn't happen again.

<details><summary>Searchable metadata</summary>

```
trac-id__6087 6087
type__release_blocker__regression release blocker: regression
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_warnings_6087 branches/warnings-6087
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1349795279000000 1349795279000000
changetime__1352145538000000 1352145538000000
version__None None
owner__itamar itamar
cc__exarkun cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [36022]) Branching to 'warnings-6087'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst removed owner|
|-|-|

I've fixed it, I think. Note that this involves a minor change in the way we do things; when we write a warning to a file in `LogPublisher.showwarning`, we use the *original* `warnings.showwarning` (or at least, what was there at import time). Previously we used the version captured by `startLoggingWithObserver`, which caused problems using `LogPublisher.showwarning` before logging had been initialized. It is vaguely possible this will break some obscure code somewhere that overrides showwarning sufficiently to actually write to files differently, and does the override in between importing `twisted.python.log` and starting twisted logging. I am inclined not to care, but I could come up with a different solution.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/warnings-6087

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.

  1. in `test_publisherWarningInitialization`
    1. the builtin `setattr` is preferable to using `warnings.__setattr__`.
  1. `twisted/python/log.py`
    1. As discussed on IRC, the `showwarning` APIs here aren't intended to be used by applications.  I think we should move towards deprecating them and not worry about all the super obscure edge cases in their behavior (related to what order events happen in, etc).

Not sure many more review comments make sense, given that hopefully the last point above means things will change significantly and for the better.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set owner to @exarkun|
|-|-|

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/warnings-6087 has started.

I implemented a simpler "fix"; we're back to the previous behavior from before the Python 3 changes, I believe, with some tests skipped or munged until they work.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @itamarst|
|-|-|

Okay.  Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst set status to `closed`|
|-|-|

(In [36349]) Merge warnings-6087.

Author: itamar
Review: exarkun
Fixes: [#6087](https://github.com/twisted/twisted/issues/6087)

The Python 3 port of logging mistakenly broke warning integration; this fixes it. It's a .misc because from point of view of upgrading user nothing will have changed.
